### ISSUE TITLE
Animal ID Types

### ISSUE THREAD

**Issue by mkm1879** [AAVLD-USAHA-ITStandards/eCVI#3]
I've pushed an edited version of ecvi2.xsd to the repository.  Instead of a choice of tag types it now has just three choices KnownFormatID, OtherFormatID, and BRAND_IMAGE.  The latter is a suggesting for incorporating those.  

I also brought back the Type attribute from version one.  I wasn't sure from yesterday's discussion if the committee finds value there.

The comments and the regular expressions definitely need work.  And this is a DRAFT for discussion at this point.  Just wanted to get a "straw dog" out there.

**Comment by mkm1879** [comment]
Recently distributed three versions of the AnimalTag element definition.  Is the additional definition provided worth the need to force change.

This needs serious discussion.

**Comment by mkm1879** [comment]
On behalf of Dee Ellis:

Man I cannot escape animal ID no matter what. Since the National Pork Board is paying for the development of our new mCVI we are obviously cognizant of their unique animal ID usage. See Patrick Webb’s email below. In thinking about what he sent me – and in light of our discussion about only validating 840 tags and NUES tags I have a few questions for y’all. 

1. When we say “validate” does it mean it validates the total # of digits,  or should it validate the exact sequence of 2 numbers 3 letters and then 4 #’s?
2. Do we then have to have a separate box for 8 digit tags vs 9 digit NUES or just have it validate as long as it is one of the 2 proper formats?
3. If it should validate proper sequence then Patrick is right – we should think about validating Prem Id’s in V3. I think LID’s aren’t as uniform – they have 6 and 8 digit sequences?

So all the other official ID ‘s mentioned below besides 840 tags and NUES we will probably have a drop down box for swine in the ”Other official (non-validated) ID box.
I don’t know about trying to correlate types of ID’s by purpose of movements.

**Comment by mkm1879** [comment]
My overly long reply:

Remember, the standard says NOTHING about how an eCVI application collects the data or what an animal health database does with the data.  It just defines the structure of those data.  Of course we try to support the widest possible range of both.  

1. “Validate” is an XML jargon term.  It is the process by which a computer program called a “parser” compares the data file with the schema file and decides if the data file follows the rules in the schema file.  We can’t encode all the “business rules” around eCVIs in the schema but we get as much as we can.  

Animal IDs have been the trickiest.  We tried checking against EVERY known official format.  I even wrote a version of the schema to do that.  It was unreadably long and complex and still failed to accept some legal IDs.  The current version is a compromise.  It lets the sending system declare that the ID is an 840 RFID, a 900 series RFID, a NUES9 or NUES8.  If it does so, the ID must match the pattern for the declared type (“No 14 digit 840 tags allowed” as Dr. Smith and I found in the form of a Civet bug.)  Any other official ID, the sending system just says, “This is another official ID” and can send anything.  “This really is an official ID” would not cause the parser to complain (“The data to fail to validate against the schema” to be correct jargon.)  Other Official ID does ask for the type, but doesn’t do validation against it.  “PIN + Management” for lack of a better phrase is a good example of an official ID format that we can’t attempt to validate because, as you say, the “local” part is so variable.  We could have put in complicated code to check that the first 7 characters look like a valid PIN and that there is something after that.  But at that point we decided it was too complicate.  (Nothing prevents an eCVI from having formatted input for specific types like that.  My Civet program lets you enter the PIN part once and just start adding the local part for the rest, for example.)

We do enumerate the PIN tags “NPIN” but probably need a distinction for the ones that also contain the management ID to make them unique.  “PINPLUS” maybe?

2. NUES9 and NUES8 are separate.  NUES8 is especially tricky because the computer can’t guess between NUES8 and Back Tag.  It would be a fairly simple user interface feature to have the user select something like “Brite tag” and then have the computer count characters.  The two together would clearly distinguish NUES8, Back Tag, and NUES9.  

3. I really would love to validate the PIN/LID field.  At this point doing the checksum calculation in schema language is beyond my capability.  Both PINs and LIDs have the same checksum calculated final character and as you say 7 (PIN) or 6 or 8 (LID) characters total.  All my Java and SQL software absolutely validates these before accepting the input.  The checksum code is pretty simple in a “real” programming language.  Schema language is so specialized it doesn’t excel in simple math.  It can just check length between 6 and 8 characters, which it does.  

I think our discussion around ID types, purpose of movement, etc., needs to focus on, “Have we collected enough information for the systems on the two ends to make decisions such as, ‘Are the IDs provided sufficient for this movement?’.”  The data standard just needs to make sure the receiving system can “understand” the information well enough to process it.  The standard doesn’t actually do that decision making.  There will be lots of CVIs that get communicated using the standard that will fail to satisfy the receiving state’s (and sending state’s) requirements.  But the data will be clear enough to allow that to be obvious and useful to things like permitting systems.
